### ISSUE TITLE
[HOLD for payment 2024-12-16] [HOLD for payment 2024-11-11] [$250] Categories -Category name for spend is not updated offline & a new category is created instead

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#50563]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.47-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+pso@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to workspace settings > Categories
3. Open Travel category (because it is linked to Airlines spend in Settings)
4. Change the category name and save it
5. Click Settings
6. Note that the category for "Airlines" is also updated
7. Go offline
8. Open Travel category, rename and save it
9. Click Settings
10. Click Airlines

## Expected Result:
Travel category will gray out and the name is updated when the category is updated offline

## Actual Result:
Travel category is not grayed out and the name is not updated
A new category with the updated name is created down in the list

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/3d2708e5-4a84-4aca-83e0-250ef594fc23

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021844365273557975639</li>
        <li>Upwork Job ID: 1844365273557975639</li>
        <li>Last Price Increase: 2024-10-17</li>
<li>Automatic offers: </li>
<ul>
<li>huult | Contributor | 104555728</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @eVoloshchak</details>

**Comment by lanitochka17** [comment]
@bfitzexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by mkzie2** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Travel category is not grayed out and the name is not updated
A new category with the updated name is created down in the list

### What is the root cause of that problem?

We haven't updated optimistic data for policy's `mccGroup`

### What changes do you think we should make in order to solve the problem?


Update optimistic data for `mccGroup` to change the corresponding category name when `renamePolicyCategory`:

```javascript
const policyCategories = Object.keys(allPolicyCategories?.[`${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${policyID}`] ?? {});
const mccGroup = policy?.mccGroup;

const updatedMccGroups: Record<string, MccGroup> = lodashCloneDeep(mccGroup ?? {});

Object.values(mccGroup ?? {}).forEach((group) => {
    if (policyCategories.includes(group.category)) {
        updatedMccGroups[group.groupID].category = policyCategory.newName;
    }
})

...
optimisticData: [
    ...,
    {
          onyxMethod: Onyx.METHOD.MERGE,
          key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
          value: {
              mccGroup: {
                  ...updatedMccGroups,
              }
          },
      },
]
```

We might need to update mccGroup's `pendingAction` as well.

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Category name for spend is not updated offline & a new category is created instead

### What is the root cause of that problem?

We do not update the category in the `mccGroup` of the `policy` when renaming the policy category.

### What changes do you think we should make in order to solve the problem?

We must update the category in the `mccGroup` of the policy when renaming the policy category. It should be something like this:

```diff
//src/libs/actions/Policy/Category.ts#L546
+  const mccGroup = policy?.mccGroup ?? [];
+  const updateMCCGroup = CategoryUtils.getCategoryMccGroup(mccGroup, policyCategory.oldName, policyCategory.newName);

//src/libs/actions/Policy/Category.ts#L641
value: {
                    rules: {
                        approvalRules,
                    },
+                    mccGroup: CategoryUtils.getCategoryMccGroup(updateMCCGroup, policyCategory.oldName, policyCategory.newName),
                },
// .src/libs/CategoryUtils.ts#L60
+ function getCategoryMccGroup(mccGroup: Record<string, MccGroup>, oldCategoryName: string, newCategoryName: string) {
+    for (const key in mccGroup) {
+        if (mccGroup[key].category === oldCategoryName) {
+            mccGroup[key].category = newCategoryName;
+        }
+    }
+    return mccGroup;
+ }
```

**Comment by bfitzexpensify** [comment]
Couple of proposals ready for review @eVoloshchak

**Comment by eVoloshchak** [comment]
@mkzie2, thank you for the [proposal](https://github.com/Expensify/App/issues/50563#issuecomment-2405145287), however, when testing it, I've encountered a couple of problems:
- The category name is not updated in Settings
- The app crashes when you click on an old category name in Settings


https://github.com/user-attachments/assets/5a5fe2ea-e70e-44d4-aa2a-f928e4778e69

**Comment by eVoloshchak** [comment]
@huult [proposal](https://github.com/Expensify/App/issues/50563#issuecomment-2405667734) looks and tests well, let's proceed with it

🎀👀🎀 C+ reviewed!

**Comment by huult** [comment]
@youssef-lr I look forward to receiving your assignment

**Comment by huult** [comment]
I will create a pull request within 24 hours. Thanks

**Comment by huult** [comment]
[PR](https://github.com/Expensify/App/pull/51390) has already been created.

**Comment by huult** [comment]
@eVoloshchak I have created a [new pull request](https://github.com/Expensify/App/pull/51927) to fix this issue and issue [#51812](https://github.com/Expensify/App/issues/51812). Could you check it? Thanks!

**Comment by huult** [comment]
@eVoloshchak Could you please review [this pull request](https://github.com/Expensify/App/pull/51927) when you have time? Thank you.

**Comment by bfitzexpensify** [comment]
Still waiting on the updated changes to be deployed here, after https://github.com/Expensify/App/issues/51812 was caused by https://github.com/Expensify/App/pull/51390

**Comment by huult** [comment]
waiting for @eVoloshchak to review the pull request

**Comment by huult** [comment]
same above

**Comment by huult** [comment]
@eVoloshchak Please review my pull request. Thank you.

**Comment by huult** [comment]
@eVoloshchak Please review my pull request. Thank you.

**Comment by huult** [comment]
@eVoloshchak Could you review the pull request so it can be completed? It has been open for 3 weeks and involves a small change, so your prompt review would be appreciated.

**Comment by huult** [comment]
Waiting for @youssef-lr to review. https://github.com/Expensify/App/pull/51927

**Comment by huult** [comment]
The pull request is already merged into the main branch

**Comment by bfitzexpensify** [comment]
Payment complete to @huult. 

@eVoloshchak, please complete the BZ checklist and I'll post the payment summary for the C+ work.

**Comment by bfitzexpensify** [comment]
Bumping this @eVoloshchak

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: there isn't a PR that has caused this, this was a functionality not implemented initially
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: I don't think an additional discussion is needed, we already have the necessary item in PR author/reviewer checklists

**Regression Test Proposal**

1. Navigate to Workspace Settings > Categories.
2. Open the Travel category (because it is linked to the Airlines spend in Settings).
3. Change the category name and click Save.
4. Click Settings and verify that the Airlines category name is updated.
5. Go offline.
6. Open the Travel category, rename it, and click Save.
7. Click Settings and then open the Airlines category.
8. Verify that the Travel category has the updated name and is grayed out.
9. 

**Do we agree 👍 or 👎**

**Comment by bfitzexpensify** [comment]
Thanks!

Payment summary:

@eVoloshchak due $250 for C+ work via ND manual request.
